plugins/logs/mask.go

```diff
@@ -0,0 +1,236 @@
+
+package logs
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/open-policy-agent/opa/util"
+)
+
+type maskOP string
+
+const (
+	maskOPRemove maskOP = "remove"
+	maskOPUpsert maskOP = "upsert"
+)
+
+type maskRule struct {
+	OP           maskOP      `json:"op"`
+	Path         string      `json:"path"`
+	Value        interface{} `json:"value"`
+	escapedParts []string
+}
+
+func (r maskRule) String() string {
+	return "/" + strings.Join(r.escapedParts, "/")
+}
+
+func newMaskRule(op maskOP, path string, val interface{}) (maskRule, error) {
+	rule := maskRule{}
+
+	if len(path) == 0 {
+		return rule, fmt.Errorf("mask must be non-empty")
+	} else if path[0] != '/' {
+		return rule, fmt.Errorf("mask must be slash-prefixed")
+	}
+
+	parts := strings.Split(path[1:], "/")
+
+	if parts[0] != "input" && parts[0] != "result" {
+		return rule, fmt.Errorf("mask prefix not allowed")
+	}
+
+	escapedParts := make([]string, len(parts))
+	for i := range parts {
+		_, err := url.QueryUnescape(parts[i])
+		if err != nil {
+			return rule, err
+		}
+
+		escapedParts[i] = url.QueryEscape(parts[i])
+	}
+
+	rule = maskRule{
+		OP:           op,
+		Path:         path,
+		Value:        val,
+		escapedParts: escapedParts,
+	}
+
+	return rule, nil
+}
+
+func (r maskRule) Mask(event *EventV1) error {
+
+	if len(r.escapedParts) == 1 {
+		switch p := r.escapedParts[0]; p {
+		case "input":
+			switch r.OP {
+			case maskOPRemove:
+				event.Input = nil
+				event.Erased = append(event.Erased, r.String())
+
+			case maskOPUpsert:
+				event.Input = &r.Value
+				event.Masked = append(event.Masked, r.String())
+
+			default:
+				return fmt.Errorf("illegal mask op value: %s", r.OP)
+			}
+
+		case "result":
+			switch r.OP {
+			case maskOPRemove:
+				event.Result = nil
+				event.Erased = append(event.Erased, r.String())
+
+			case maskOPUpsert:
+				event.Result = &r.Value
+				event.Masked = append(event.Masked, r.String())
+
+			default:
+				return fmt.Errorf("illegal mask op value: %s", r.OP)
+			}
+		default:
+			return fmt.Errorf("illegal path value value: %s", p)
+		}
+	} else {
+		var node interface{}
+
+		switch p := r.escapedParts[0]; p {
+		case "input":
+			if event.Input != nil {
+				node = *event.Input
+			}
+		case "result":
+			if event.Result != nil {
+				node = *event.Result
+			}
+		default:
+			return fmt.Errorf("illegal path value value: %s", p)
+		}
+
+		parent := r.lookup(r.escapedParts[1:len(r.escapedParts)-1], node)
+		parentObj, ok := parent.(map[string]interface{})
+		if !ok {
+			return nil
+		}
+
+		fld := r.escapedParts[len(r.escapedParts)-1]
+		_, fldOK := parentObj[fld]
+
+		switch r.OP {
+		case maskOPRemove:
+			if !fldOK {
+				return nil
+			}
+			delete(parentObj, fld)
+			event.Erased = append(event.Erased, r.String())
+
+		case maskOPUpsert:
+
+			inputObj, ok := node.(map[string]interface{})
+			if !ok {
+				return nil
+			}
+
+			delete(inputObj, fld)
```

Comment:
Is the delete statement needed if we are overwriting below ?

Reply:
It is not, will remove.